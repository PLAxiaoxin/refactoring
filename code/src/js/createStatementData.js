export function createStatementData(invoice, plays) {
  class PerformanceCalculator {
    constructor(aPerformance, aPlay) {
      this.performance = aPerformance;
      this.play = aPlay;
    }

    get amount() {
      throw new Error('subclass responseibility');
    }

    get volumeCredits() {
      return Math.max(this.performance.audience - 30, 0);
    }
  }

  class TragedyCalculator extends PerformanceCalculator {
    get amount() {
      let result = 40000;
      if (this.performance.audience > 30) {
        result += 1000 * (this.performance.audience - 30);
      }
      return result;
    }
  }

  class ComdyCalcuator extends PerformanceCalculator {
    get amount() {
      let result = 30000;
      if (this.performance.audience > 20) {
        result += 10000 + 500 * (this.performance.audience - 20);
      }
      result += 300 * this.performance.audience;
      return result;
    }

    get volumeCredits() {
      return super.volumeCredits + Math.floor(this.performance.audience / 5);
    }
  }

  const result = {};
  result.customer = invoice.customer;
  // 通过数组的 map 方法，为invoice.performances 中的每一项，获取匹配的 play
  result.performances = invoice.performances.map(enrichPerformance);
  result.totalAcount = totalAcount(result);
  result.totalVolumeCredits = totalVolumeCredits(result);
  return result;

  function enrichPerformance(aPerformance) {
    const calculator = ceratePerformanceCalculator(
      aPerformance,
      playFor(aPerformance),
    );
    const result = Object.assign({}, aPerformance);
    result.play = calculator.play;
    result.amount = calculator.amount;
    result.volumeCredits = calculator.volumeCredits;
    return result;
  }

  function playFor(aPerformance) {
    return plays[aPerformance.playID];
  }

  function totalAcount(data) {
    return data.performances.reduce((total, p) => total + p.amount, 0);
  }

  function totalVolumeCredits(data) {
    return data.performances.reduce((total, p) => total + p.volumeCredits, 0);
  }

  // 以多态取代条件表达式
  function ceratePerformanceCalculator(aPerformance, aPlay) {
    switch (aPlay.type) {
      case 'tragedy':
        return new TragedyCalculator(aPerformance, aPlay);
      case 'comedy':
        return new ComdyCalcuator(aPerformance, aPlay);
      default:
        throw new Error(`unkown type: ${aPlay.type}`);
    }
  }
}
