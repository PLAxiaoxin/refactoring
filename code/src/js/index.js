import { invoices } from '../data/invoices.js';
import { plays } from '../data/plays.js';
import { createStatementData } from './createStatementData.js';
console.log(statement(invoices, plays));
function statement(invoice, plays) {
  return renderPlanText(createStatementData(invoice, plays));
}

function renderPlanText(data) {
  let result = `Statement for ${data.customer}\n`;

  for (let pref of data.performances) {
    result += `${pref.play.name}: ${usd(pref.amount)} (${
      pref.audience
    } seats)\n`;
  }

  result += `Amount owed is ${usd(data.totalAcount)}\n`;
  result += `You earned ${data.totalVolumeCredits} credits\n`;
  return result;
}

function htmlStatement(invoices, plays) {
  return renderHtml(createStatementData(invoice, plays));
}

function renderHtml(data) {
  let result = `<h1>Statement for ${data.customer}</h1>`;
  result += '<Table>\n';
  result += '<tr><th>play</th><th>seats</th><th>cost</th></tr>';
  for (let pref of data.preformance) {
    result += `<tr><td>${pref.play.name}</td><td>${pref.audience}</td>`;
    result += `<td>${usd(pref.amount)}</td></tr>/n`;
  }
  result += '</table>/n';
  result += `<p>Amount owend is <em>${usd(data.totolAmount)}</em></p>/n`;
  result += `<p>You earned <em>${usd(data.totolVolumeCredits)}</em></p>/n`;
}

function usd(aNumber) {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2,
  }).format(aNumber / 100);
}
