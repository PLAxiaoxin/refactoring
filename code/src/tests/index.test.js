import { describe, it, expect } from 'vitest';
import { invoices } from '../data/invoices.js';
import { plays } from '../data/plays.js';
import { textStatement, htmlStatement, usd } from '../js/index.js';

describe('happy path', () => {
  it('output text', () => {
    expect(textStatement(invoices, plays)).toMatchSnapshot();
  });

  it('output html', () => {
    expect(htmlStatement(invoices, plays)).toMatchSnapshot();
  });

  it('test util', () => {
    const text = usd(1000);
    expect(text).toBe('$10.00');
  });
});
