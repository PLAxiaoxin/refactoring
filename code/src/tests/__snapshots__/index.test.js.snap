// Vitest Snapshot v1

exports[`happy path > output html 1`] = `
"<h1>Statement for BigCo</h1><Table>
<tr><th>play</th><th>seats</th><th>cost</th></tr><tr><td>Hamlet</td><td>55</td><td>$650.00</td></tr>/n<tr><td>As You Like It</td><td>35</td><td>$580.00</td></tr>/n<tr><td>Othello</td><td>40</td><td>$500.00</td></tr>/n</table>/n<p>Amount owend is <em>$NaN</em></p>/n<p>You earned <em>$NaN</em></p>/n"
`;

exports[`happy path > output text 1`] = `
"Statement for BigCo
Hamlet: $650.00 (55 seats)
As You Like It: $580.00 (35 seats)
Othello: $500.00 (40 seats)
Amount owed is $1,730.00
You earned 47 credits
"
`;
